src/auth/index.ts

```diff
@@ -0,0 +1,683 @@
+    where: (builder) =>
+      builder.whereIn(
+        'id',
+        grants.map((g) => g.target)
+      ),
+  });
+  const targetsById = organizeObjectsByUniqueProperty(targets, 'id');
+
+  const result = new Map<ParticipantId, RolesGrant[]>();
+
+  for (const grant of grants) {
+    const grantee = granteesById.get(grant.grantee);
+    if (!grantee) {
+      throw new Error(`Unexpected missing grantee`);
+    }
+
+    const target = targetsById.get(grant.target);
+    if (!target) {
+      throw new Error('Unexpected missing target');
+    }
+
+    const participantId: ParticipantId = createBrandedValue(grantee.granteeId);
+    let list = result.get(participantId);
+    if (!list) {
+      list = [];
+      result.set(participantId, list);
+    }
+
+    list.push(
+      calculateRolesGrantFromTargetAndRoleStrings(
+        target,
+        grant.roles,
+        logContext
+      )
+    );
+  }
+
+  return result;
+};
+
+export const getRoleGrantsForUser = async ({
+  context,
+  user,
+  logContext,
+}: {
+  context: Pick<Context, 'connection'>;
+  user: ParticipantId;
+  logContext: LogContext;
+}): Promise<RolesGrant[]> =>
+  (
+    await getRoleGrantsForUsers({
+      context,
+      users: [user],
+      logContext,
+    })
+  ).get(user) || [];
+
+/**
+ * Calculate the complete set of permissions for the actor of the current req
+ *
+ * TODO: cache this calculation, and also store it in the request object
+ */
+export const calculatePermissions = async (
+  context: Context
+): Promise<perms.GrantedPermissions> => {
+  const allowed: perms.GrantedPermissions = {};
+
+  const logContext = await createRootContextFromRequest(context);
+
+  // Both legacy and new permissions require us to know the current participant
+  // rather than require both simultaneously race to calculate this,
+  // calculate it once here (as it's cached)
+  const participant = await getLoggedInParticipant(context, logContext);
+  if (!participant) {
+    return allowed;
+  }
+
+  const mergeAllowedPermissions = <
+    Type extends keyof Omit<perms.GrantedPermissions, 'global'>
+  >(
+    type: Type,
+    additions: Map<number, Set<perms.PermissionStrings<Type>>>
+  ) => {
+    let existingMap = allowed[type] as
+      | Map<number, Set<perms.PermissionStrings<Type>>>
+      | undefined;
+    if (!existingMap) {
+      existingMap = allowed[type] = new Map();
+    }
+    for (const [key, permissions] of additions.entries()) {
+      let set = existingMap.get(key);
+      if (!set) {
+        existingMap.set(key, (set = new Set()));
+      }
+      for (const p of permissions) {
+        set.add(p);
+      }
+    }
+  };
+
+  const addLegacyPermissions = async () => {
+    const globallyPermitted = await getLegacyGloballyPermittedActions(context);
+    if (globallyPermitted.size === 0) {
+      return;
+    }
+    allowed.global = allowed.global || new Set();
+    for (const permission of globallyPermitted) {
+      allowed.global.add(permission);
+    }
+  };
```

Comment:
this we will want to keep in `hpc_service` only, to support the endpoints that haven't been updated to use the new auth system only.